### ISSUE TITLE
"No PSR-17 request factory found" exception

### ISSUE THREAD

**Issue by mitchhentges** [getsentry/sentry-php#988]
### What's going on :thinking: :thinking: 

When calling `ClientBuilder->getClient()`, it would fail because:
> `No PSR-17 request factory found. Install a package from this list: https://packagist.org/providers/psr/http-factory-implementation`

This is thrown all the way down in `php-http/discovery` [here](https://github.com/php-http/discovery/blob/master/src/Psr17FactoryDiscovery.php#L22), which is failing because it can't find a library to do HTTP with.

There's two reasons for this:
1. `sentry-php` changed to use `guzzlehttp/psr7` instead of `zendframework/zend-diactoros` in [`v2.3.0`](https://github.com/getsentry/sentry-php/releases/tag/2.3.0)
> Replace dependency to zendframework/zend-diactoros with guzzlehttp/psr7 (#945)
2. `php-http/discovery` can't actually discover `guzzlehttp/psr7` (their issue reported [here](https://github.com/php-http/discovery/issues/164)).

### How to fix :) :)

weaverryan has pointed out in the [`php-http/discovery` ticket](https://github.com/php-http/discovery/issues/164) that an interop library for `guzzle` exists called `http-interop/http-factory-guzzle`. By depending on that library instead of `guzzle` directly, then `discovery` will work and cheers will be had.
**However**, the interop library hasn't been updated in a hot minute, so perhaps Sentry would want to define the interop itself so it can use a newer version of `guzzle`? Perhaps you'd like to ping upstream to get them to update? I'm not sure, you call the shot :)

### TL;DR:

Since Sentry 2.3.0, Sentry fails to run unless Sentry users happen to install the right HTTP library.
To solve this, depend on `http-interop/http-factory-guzzle` or implement its behaviour within `sentry-php`

**Comment by ste93cry** [comment]
[`sentry-php-sdk`](https://github.com/getsentry/sentry-php-sdk) already requires the `http-interop/http-factory-guzzle` package in its `composer.json`. This package instead contains the source code of the core SDK, and should not force any implementation of an HTTP client because it's up to you to decide which you want to use: if you don't want to make this choice then the first package is the right one to pick.

**Comment by mitchhentges** [comment]
Makes sense to me!

**Comment by ste93cry** [comment]
After re-reading with more attention your first comment and the referenced issue in the `php-http/discovery` repository I now see the root cause of the problem and I think that it should be indeed fixed upstream. I will try to open a PR to get this fixed and see how it goes. I also just noticed that `guzzlehttp/psr7` is exposing to Composer that it provides an implementation for `psr/http-factory-implementation`, thus making `http-interop/http-factory-guzzle` useless, but this change has not been released yet and this is why you get that error. It used to work in the past because `zendframework/zend-diactoros` did this from the beginning, so no extra package to make it working was needed
